vignettes/qgisprocess.Rmd

```diff
@@ -0,0 +1,271 @@
+library("terra")
+# attach digital elevation model from Mt. Mongón (Peru)
+dem <- rast(system.file("raster/dem.tif", package = "spDataLarge"))
+
+info <- qgis_run_algorithm(alg = "grass7:r.slope.aspect", elevation = dem)
+                           
+```
+
+Just printing the `info` object shows which output objects have been made:
+
+```{r info}
+info
+```
+
+Combine these output rasters as a multi-layered `SpatRaster` object and plot it:
+
+```{r combine1, message=FALSE}
+# just keep the names of output rasters
+nms <- qgis_outputs("grass7:r.slope.aspect")$name
+# read in the output rasters 
+r <- info[nms] |>
+  unlist() |>
+  terra::rast() |>
+  as.numeric()
+names(r) <- nms
+# plot the output
+plot(r)
+```
+
+An alternative way to combine the rasters is given below.
+
+```{r combine2, message=FALSE}
+r <- lapply(info[nms], \(x) as.numeric(qgis_as_terra(x))) |> 
+  rast()
+plot(r)
+```
+
+## Piping
+
+`qgis_process` does not lend itself naturally to piping because its first argument is the name of a geoalgorithm instead of a data object.
+`qgis_pipe()` circumvents this by accepting a `.data` object as its first argument, and pipes this data object into the the first argument of a geoalgorithm assuming that the specified geoalgorithm in fact needs a data input object as its first argument.
+
+```{r buf-pipe}
+system.file("longlake/longlake_depth.gpkg", package = "qgisprocess") |>
+  qgis_pipe("native:buffer", DISTANCE = 50)
+```
+
+In case of a `qgis_result` object, `qgis_pipe()` automatically tries to select an element named `OUTPUT`.
+However, if the output has another name (e.g., `DEM_PREPROC` as in the example below) or if there are multiple output elements to choose from (e.g., `sagang:sagawetnessindex` has four output rasters, check with `qgis_outputs("sagang:sagawetnessindex")`, you can specify the wanted output object via the `.select` argument.
+
+```{r qgis-pipe}
+dem <- system.file("raster/dem.tif", package = "spDataLarge")
+# in case you need to enable the SAGA next generation algorithms, run the following lines:
+# qgis_plugins()
+# qgis_enable_plugins("processing_saga_nextgen")
+qgis_run_algorithm(algorithm = "sagang:sinkremoval", DEM = dem, 
+                   METHOD = 1) |>
+  qgis_pipe("sagang:sagawetnessindex", .select = "DEM_PREPROC")
+```
+
+When piping, `qgis_pipe()` automatically cleans up after you by deleting intermediate results.
+This avoids cluttering your system when running geoalgorithms on large spatial data files.
+To turn of this behavior, set `.clean` to `FALSE`.
+
+Of course, you can also use the native pipe.
+This avoids ambiguity and allows for greater flexibility though it might not be as convenient as `qgis_pipe()`.
+For example, intermediate results remain on disk for the duration of your R session.
+
+```{r native pipe}
+out = qgis_run_algorithm(algorithm = "sagang:sinkremoval", DEM = dem, 
+                         METHOD = 1) |>
+   (\(x) qgis_run_algorithm(algorithm = "sagang:sagawetnessindex",
```

Comment:
FWIW I doubt whether the lambda syntax as shorthand for anonymous functions is well known among users (it's only there since R 4.1), but of course it's valid, but it may distract from the point you are making. (The placeholder syntax is simpler)

Reply:
yes, I plan to use the placeholder as soon as R 4.3 is available on Ubuntu 22.04

Reply:
4.3.0 is available, if one follows the R installation instructions for Ubuntu at https://cran.r-project.org/bin/linux/ubuntu/. The regular Ubuntu repo does not follow R versions this quickly, if at all.

```bash
$ apt policy r-base
r-base:
  Installed: 4.3.0-1.2204.0
  Candidate: 4.3.0-1.2204.0
  Version table:
 *** 4.3.0-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
        100 /var/lib/dpkg/status
     4.2.3-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.2.20221110-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.2-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.1-3.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.1-2.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.1-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.0-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.1.2-1ubuntu2 500
        500 http://ftp.belnet.be/ubuntu jammy/universe amd64 Packages
        500 http://ftp.belnet.be/ubuntu jammy/universe i386 Packages
```

Reply:
sorry, I have been unclear. Of course, you are right, that we would have to wait forever for the latest R version using the regular Ubuntu repo. I was referring to the [rocker/geospatial Dockerfiles](https://github.com/rocker-org/rocker-versioned2/pkgs/container/geospatial/versions) on which the ghcr.io/geocompx/docker:qgis image is based. But in fact, the geospatial images already come with R >4.3. When will this come into effect in the geocompx images, @Robinlovelace?